### PR TITLE
[SPARK-32994][CORE] Update external heavy accumulators before they entering into listener event loop

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?
Add a new accumulator prefix `external.heavy.` to determine the heavy accumulators. These accumulators will be handled before entering into `eventProcessLoop`. This could avoid the driver holding the heavy accumulators in the event loop for a long time. Then the driver memory could be release in young GC.

### Why are the changes needed?
We use Spark + Delta Lake, recently we find our Spark driver faced full GC problem (very heavy) when users submit a MERGE INTO query. The driver held over 100GB memory (depends on how much the max heap size set) and can not be GC forever. By making a heap dump we found the root cause. More details about the issue and the patch testing result, see [SPARK-32994](https://issues.apache.org/jira/browse/SPARK-32994)

Screen Shot 2020-09-25 at 11 35 01 AM

From above heap dump, Delta uses a SetAccumulator to records touched files names
```scala
    // Accumulator to collect all the distinct touched files
    val touchedFilesAccum = new SetAccumulator[String]()
    spark.sparkContext.register(touchedFilesAccum, TOUCHED_FILES_ACCUM_NAME)

    // UDFs to records touched files names and add them to the accumulator
    val recordTouchedFileName = udf { (fileName: String) => {
      touchedFilesAccum.add(fileName)
      1
    }}.asNondeterministic()
```

In a big query, each task may hold thousands of file names, and if a stage contains dozens of thousands of tasks, DAGscheduler may hold millions of `CompletionEvent`. And each `CompletionEvent` holds the thousands of file names in its `accumUpdates`. All accumulator objects will use Spark listener event to deliver to the event loop and even a full GC can not release memory.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a UT

### CODE REVIEW

**Comment by LantaoJin** [comment]
ping @cloud-fan @HyukjinKwon @dongjoon-hyun

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -274,8 +274,36 @@ private[spark] class DAGScheduler(
       accumUpdates: Seq[AccumulatorV2[_, _]],
       metricPeaks: Array[Long],
       taskInfo: TaskInfo): Unit = {
+    // It may cause heavy full GC problem if some heavy external accumulators keep in memory.
+    // We update these heavy accumulators before they entering into Spark listener event loop.
+    val (heavyAccumUpdates, otherAccumUpdates) = accumUpdates.partition { acc =>
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:279]: "Heavy" means that accumulated values are too big? If so, we cannot just check the size here and pre-merge it if the size goes over a threshold? IMHO it is not a good idea for developers to control the accumulator behavior explicitly via the name prefix.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -274,8 +274,36 @@ private[spark] class DAGScheduler(
...
+      acc.name.exists(_.startsWith(ExternalHeavyAccumulator.HEAVY_PREFIX))
+    }
+    heavyAccumUpdates.foreach { updates =>
+      val id = updates.id
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:283]: This code part looks almost the same with `updateAccumulators`, so could we share it between them?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:279]: Yes. "Heavy" means the size is too big, such as the accumulator value contains thousands of file paths. I know this looks not a good solution. Do you mean to use a size threshold to filter out the big accumulator values is a better solution? The problem that way is we split the same accumulators (same name with different sizes). Actually, Spark uses METRICS_PREFIX to distinguish the internal metrics from all accumulators `val METRICS_PREFIX = "internal.metrics."`. We can split to `val (externalAccumUpdates, internalAccumUpdates)` and update/merge the `externalAccumUpdates` before they entering into event loop. But, you know, we still cannot forbid developers to use the "internal.metrics." prefix. We only could add some comments in code and document it.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 2 above)
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:283]: Yes. We can extract the common part as a new private method.

**Comment by tgravescs** [comment]
I haven't looked at this in detail but you say there is no user facing changes but I don't think that's true. If a user names a accumulator in a certain way then it does something different correct? That is user impacting api and that would need to be documented somewhere. I'm personally not a fan of the word "heavy" here as I think its not clear as to what this means or consequences of it.  I would also consider this a new feature not a bug fix since it introduces new api

**Comment by LantaoJin** [comment]
I can add something in the user facing changes part. But let me explain why I written No from another viewpoint. This patch actually separates all accumulators to two types: merged before entering event loop and merged after entering event loop. Without this patch all accumulators are the second type. With this patch, whatever name user set, if accumulators with a certain name go to the first type, they will benefited by this patch, avoid triggering fullGC. If their with another certain name go to the second type, these accumulators behivous are just like without patch, nothing will go worse.
Of course, change a certain name to match the first type will be good. but this patch is just give an opportunity for developers to avoid driver full GC. They can keep their old accumulator name if the driver are good enough. As long as this patch not make things worse, I call that no user facing changed. Of course, it is not a bug, it’s should be a improvement or feature. if this approach acknowledged, I will document it somewhere.

**Comment by LantaoJin** [comment]
BTW, “heavy” is a simple work for me, a non-native English speaker. Maybe it could be removed and change the title to “Add a new interface (metrics prefix) to merge accumulators before entering event loop”

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:279]: internal configs and user facing are completely different.  Yes if someone reads the code they could create internal ones but that is use at your own risk and we don't guarantee any api. This is user facing api essentially.  I agree with @maropu I'm not really fond of the prefix approach.   Is this only for CollectionAccumulator or SetAccumulator? Although it looks like the SetAccumulator is specific to delta.

**Comment by tgravescs** [comment]
thanks for the explanation but that isn't the intention of that field.  Not making things worse is great but doesn't indicate whether it is user facing. I can add a completely new api to RDD that if not used doesn't "break" anything but its adding a user facing API and extra functionality that needs to be documented so the user knows its there and can use it. this "api" change is meant for end users to use, so how do they know its there to use - most users don't go read the code to find little options like this.  this is adding a somewhat obscure special prefix which does special things. I'm not sure I really like that approach. The nice thing about it is it doesn't require any function API changes, but that also makes it not as obvious.

**Comment by LantaoJin** [comment]
Thanks @tgravescs , actually I agree with your comment. I just told you what I thought. Current code add a user facing changes in deed. I know current patch is not good. Actually to fix this problem I hardcode to filter out the delta SetAccumulator by name in our internal Spark. Open this PR is to help expose the issue and find a common solution. Now looks like there are four ways:
1. Add a new prefix like “heavy” to merge them before entering loop. (Current patch, and no one like it)
2. Merge accumulators which not start with  MATRICS_PREFIX before entering loop and document it.
3. Merge accum which size over a threshold before entering ...
4. Merge the collection type accum before entering ...

More comments.

**Comment by maropu** [comment]
> Do you mean to use a size threshold to filter out the big accumulator values? The problem that way is we split the same accumulators (same name with different sizes). 

Yea, I meant so. What did `split` mean in this scenario?

**Comment by cloud-fan** [comment]
Will this PR slow down the DAGScheduler? The accumulator handling was async before.

**Pull request closed without merging**